packages/react-native-gesture-handler/src/v3/NativeDetector/HostGestureDetector.web.tsx

```diff
@@ -72,12 +87,55 @@ const HostGestureDetector = (props: GestureHandlerDetectorProps) => {
       );
     }
 
-    attachHandlers(new Set(handlerTags));
+    const currentHandlerTags = new Set(handlerTags);
+    detachHandlers(currentHandlerTags, attachedHandlers.current);
+
+    attachHandlers(
+      viewRef,
+      propsRef,
+      currentHandlerTags,
+      attachedHandlers.current,
+      ActionType.NATIVE_DETECTOR
+    );
   }, [handlerTags, children]);
 
+  useEffect(() => {
+    const logicChildrenToDetach: Set<number> = new Set(
+      attachedLogicHandlers.current.keys()
+    );
+
+    props.logicChildren?.forEach((child) => {
+      if (!attachedLogicHandlers.current.has(child.viewTag)) {
+        attachedLogicHandlers.current.set(child.viewTag, new Set());
+      }
+      logicChildrenToDetach.delete(child.viewTag);
+
+      const currentHandlerTags = new Set(child.handlerTags);
+      detachHandlers(
+        currentHandlerTags,
+        attachedLogicHandlers.current.get(child.viewTag)!
+      );
+
+      attachHandlers(
+        child.viewRef,
+        propsRef,
+        currentHandlerTags,
+        attachedLogicHandlers.current.get(child.viewTag)!,
+        ActionType.LOGIC_DETECTOR
+      );
+    });
+
+    logicChildrenToDetach.forEach((tag) => {
+      detachHandlers(new Set(), attachedLogicHandlers.current.get(tag)!);
+    });
+  }, [props.logicChildren]);
+
   useEffect(() => {
     return () => {
-      detachHandlers(attachedHandlerTags.current);
+      detachHandlers(new Set(), attachedHandlers.current);
```

Comment:
Maybe it won't change much in the optimization, but we can create `EMPTY_SET` variable and then pass it to avoid calling these constructors.